### PR TITLE
WIP: Fix preferences window not appearing when the content views are not using auto-layout

### LINKED ISSUE #9: Rapidly switching between tabs causes view size to be wrong
prefs

Is there an option to force view size recalculation every time another tab is selected?

### PR DESCRIPTION
**Pull request by SamusAranX**
This is an attempt to fix #9, specifically the part where the preference window doesn't show up.

I tried to fix this by using `viewController.view.bounds.size` in cases where `viewController.view.fittingSize` is `.zero` and caching the resulting value. 
Without the cache, both `bounds.size` and `fittingSize` become `.zero` later on, which would cause problems.

### CODE REVIEW

**Comment by SamusAranX** [comment]
Oh yeah, this also means that users don't have to set `self.preferredContentSize` themselves. I previously suggested this as a band-aid in https://github.com/sindresorhus/Preferences/issues/9#issuecomment-483565894.

**Comment by SamusAranX** [comment]
Merging this PR should probably have to wait until the #26 has been merged. This PR still has some issues that could be fixed more easily then.

**Comment by DivineDominion** [comment]
Looking forward to testing this when the others are merged :)

File: Sources/Preferences/PreferencePane.swift [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import Cocoa
 
-public struct PreferencePaneIdentifier: Equatable, RawRepresentable {
+public struct PreferencePaneIdentifier: Equatable, RawRepresentable, Hashable {
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencePane.swift]: `Hashable` inherits from `Equatable`, so you don't need both.

```suggestion
public struct PreferencePaneIdentifier: Hashable, RawRepresentable {
```

File: Sources/Preferences/PreferencesTabViewController.swift [hunk 2]
```diff
@@ -14,6 +14,8 @@ final class PreferencesTabViewController: NSViewController, PreferencesStyleCont
 		return preferencesStyleController?.toolbarItemIdentifiers() ?? []
 	}
 
+	private var viewSizeCache: [PreferencePaneIdentifier: NSSize] = [:]
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencesTabViewController.swift]:
```suggestion
	private var viewSizeCache = [PreferencePaneIdentifier: CGSize]()
```

File: Sources/Preferences/PreferencesTabViewController.swift [hunk 3]
```diff
@@ -177,11 +179,22 @@ final class PreferencesTabViewController: NSViewController, PreferencesStyleCont
...
 	private func setWindowFrame(for viewController: NSViewController, animated: Bool = false) {
-		guard let window = window else {
+		guard let window = window,
+			  let preferencePane = preferencePanes.first(where: { $0.viewController == viewController })
+			else {
 			preconditionFailure()
 		}
 
-		let contentSize = viewController.view.fittingSize
+		var contentSize = NSSize.zero
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencesTabViewController.swift]:
```suggestion
		var contentSize = CGSize.zero
```

File: Sources/Preferences/PreferencesTabViewController.swift [hunk 4]
```diff
@@ -177,11 +179,22 @@ final class PreferencesTabViewController: NSViewController, PreferencesStyleCont
 	}
 
 	private func setWindowFrame(for viewController: NSViewController, animated: Bool = false) {
-		guard let window = window else {
+		guard let window = window,
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencesTabViewController.swift]:
```suggestion
		guard
			let window = window,
```

File: Sources/Preferences/PreferencesTabViewController.swift [hunk 5]
```diff
@@ -177,11 +179,22 @@ final class PreferencesTabViewController: NSViewController, PreferencesStyleCont
...
+			  let preferencePane = preferencePanes.first(where: { $0.viewController == viewController })
+			else {
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencesTabViewController.swift]:
```suggestion
		else {
```

**Comment by sindresorhus** [comment]
> Merging this PR should probably have to wait until #26 has been merged.

#26 is merged.

**Comment by SamusAranX** [comment]
~Phew, this took way too long.~
I think I fixed all the remaining issues.
In my local testing, everything works with the exception of a small animation glitch:
animation
That might just be my weird Auto Layout setup though, so YMMV.

**Comment by sindresorhus** [comment]
Using this PR, if I disable auto-layout in interface builder for `GeneralPreferenceViewController.xib` and `AdvancedPreferenceViewController.xib`, it now correctly show the preferences window, and switching between view work. However, it no longer animates in the view. Now it just shows the views directly with no animation.

**Comment by SamusAranX** [comment]
Another thing I noticed is that the "Switch preference style" button will cause the example app to crash without writing anything to the error log. Any idea why that could happen?

**Comment by SamusAranX** [comment]
Like with most other things in Xcode, deleting the repository and re-cloning it got rid of that error. 🤷‍♂ 

On my end, the animations are missing as well, regardless of whether Auto Layout is disabled or not. Additionally, when changing tabs for the first time after opening the window, the disappearing view controller slides out of the window: 
animationglitch
I'm not sure what happens here. 

Might I suggest removing animations altogether? While they look and feel nice when they work, they also seem to break in interesting ways in situations where the layout of views and view controllers is modified directly. And [while you're right that there's no unified standard for preference windows](https://github.com/sindresorhus/Preferences/issues/15#issuecomment-479424620), the behavior of the animation-less example window matches that of Finder and Safari's preference windows.

**Comment by DivineDominion** [comment]
That's an interesting crash, and weird app behavior. Which macOS version are you on? I remember that I disabled the `animated` flag for non-Mojave build because the calculations were off even worse.

**Comment by SamusAranX** [comment]
I'm on 10.14.5. I don't think the app crash was actually related to any code in this repository, my Xcode project just somehow got mangled. Deleting and re-cloning it fixed it.

**Comment by DivineDominion** [comment]
Ok. So the latest GIF you posted is how things look consistently on your machine now? 

> Might I suggest removing animations altogether? While they look and feel nice when they work, they also seem to break in interesting ways in situations where the layout of views and view controllers is modified directly. And while you're right that there's no unified standard for preference windows, the behavior of the animation-less example window matches that of Finder and Safari's preference windows.

Your GIF's look matches what I see in e.g. Finder preferences: the window frame animates, the contents just appear suddenly. Do you want to suggest a change to this, or is this already the desired behavior and you suggest getting rid of the setting?

**Comment by SamusAranX** [comment]
Yeah, it's how things look on my machine now. That's with the changes I made though, the version currently available in this repository will probably look different.

I haven't removed animations in my version yet. The animations breaking are just a side effect of my code changes, hence why I suggested removing them. Attempting to fight Auto Layout to get animations will only lead to pain.

Another thing that's probably hard to spot: When the GIF's animation restarts and I switch away from the Advanced tab, you can see the "Advanced" label animate downwards out of the window frame for a split second. That's another animation bug.

**Comment by sindresorhus** [comment]
> Your GIF's look matches what I see in e.g. Finder preferences: the window frame animates, the contents just appear suddenly. Do you want to suggest a change to this, or is this already the desired behavior and you suggest getting rid of the setting?

I agree. That's how it should work.

In master:
- `animate: true` - the window animates the resize and the content also fade-slides in from the bottom.

What we should change it to:
- `animate: true` - the window animates the resize and the content just appears after the window resize animation is done.

@SamusAranX Can you remove the content fade-slide animation? That should fix all the current issues, right?

**Comment by SamusAranX** [comment]
It *should* fix the issues, but I don't even know how I managed to make the example project behave like it did in the GIF in the first place 😅
When I opened and ran the example project just now, I couldn't reproduce the behavior.

Also, I can't get the two view controllers to hide themselves during the resizing animation. I'm not that familiar with the animation system on macOS.

**Comment by DivineDominion** [comment]
Just an idea to toy with: maybe it helps to resize the window and in the completion handler of the window resize animation replace the view controllers, but not earlier.

**Comment by sindresorhus** [comment]
> Just an idea to toy with: maybe it helps to resize the window and in the completion handler of the window resize animation replace the view controllers, but not earlier.

@SamusAranX Could you try that?

**Comment by SamusAranX** [comment]
It's the final few exam weeks where I am, so I'm kind of out of time 😅 
I'll get to it as soon as I can though!

**Comment by SamusAranX** [comment]
I finally found the time to mess around some more. Now (as of commit 90e400f in my fork), the preferences window works as I expect it to:
tabs

The latest merge of master into my branch breaks it though, because `PreferencePane`'s `viewController` property was removed. I was relying on that to cache window measurements.

**Comment by sindresorhus** [comment]
> because PreferencePane's viewController property was removed

It was removed as it's moot. The `PreferencePane` protocol is now [guaranteed to be a `NSViewController`](https://github.com/sindresorhus/Preferences/blob/b999a0a0845afbc2a5d30c825c2e4d788a4b869e/Sources/Preferences/PreferencePane.swift#L13). So instead of using `.viewController`, you can just use `self`.

File: Sources/Preferences/PreferencePane.swift [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import Cocoa
 
-public struct PreferencePaneIdentifier: Hashable, RawRepresentable, Codable {
+public struct PreferencePaneIdentifier: Hashable, RawRepresentable {
```
Comment (sindresorhus) [inline Sources/Preferences/PreferencePane.swift]: ?

Comment (SamusAranX) [reply Sources/Preferences/PreferencePane.swift]: …huh. This might've been an accident that happened when I merged the master branch into my branch.

**Comment by SamusAranX** [comment]
Sorry for the force push, I messed up my commits halfway through and had to get rid of them.

Commit da9ebc1 attempts to fix view controller opacity animations by explicitly setting `.wantsLayer = true` on their views so the optional chain right afterwards always has a layer to work with. Instead of trying to find a completion handler that runs after the window frame animation ends, I just used `DispatchQueue.main.asyncAfter` with a delay of 250ms, since the animation takes roughly that long. In practice, there's no visible difference.

I also tried making it so that enabling Reduce Motion in the Accessibility panel disables the animation altogether, but I couldn't find a way to query the state of that setting programmatically.
A project for another day, perhaps.

**Comment by DivineDominion** [comment]
In the example app, I think the content animation makes sense the way it looks with your changes. Can you check on the segmented control implementation, though? Its position animates in a very jumpy way now.

Screen captures here: https://imgur.com/a/LP34fv1

**Comment by SamusAranX** [comment]
I've noticed that even with the regular toolbar style. It seems the toolbar's width isn't updated continuously while animating. Honestly, I'm not sure how much of this is in the code and how much of it is the OS's doing.

**Comment by lwouis** [comment]
Hi everyone! It seems this PR was going well but then activity stopped. Is this work abandoned? Cheers!

**Comment by sindresorhus** [comment]
Closing as this is not moving forward and the PR fork has been deleted.

**Pull request closed without merging**